app/plugin_test.go

```diff
@@ -592,6 +592,180 @@ func TestPluginSync(t *testing.T) {
 			require.Nil(t, err)
 			require.Len(t, pluginStatus, 1)
 			require.Equal(t, pluginStatus[0].PluginId, "testplugin")
+
+			appErr = th.App.RemovePlugin("testplugin")
+			checkNoError(t, appErr)
+
+			appErr = th.App.RemoveFile(th.App.getBundleStorePath("testplugin.tar.gz"))
+			checkNoError(t, appErr)
+			appErr = th.App.RemoveFile(fmt.Sprintf("%s.sig", th.App.getBundleStorePath("testplugin")))
+			checkNoError(t, appErr)
 		})
 	}
 }
+
+func TestCanaryPlugins(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+	os.MkdirAll("./test-plugins", os.ModePerm)
+	defer os.RemoveAll("./test-plugins")
+
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.PluginSettings.Enable = true
+		*cfg.PluginSettings.Directory = "./test-plugins"
+		*cfg.PluginSettings.ClientDirectory = "./test-client-plugins"
+		*cfg.PluginSettings.RequirePluginSignature = false
+		*cfg.PluginSettings.CanaryTag = "daily"
+	})
+
+	env, err := plugin.NewEnvironment(th.App.NewPluginAPI, "./test-plugins", "./test-client-plugins", th.App.Log)
+	require.NoError(t, err)
+	th.App.SetPluginsEnvironment(env)
+	// New bundle in the file store case
+	path, _ := fileutils.FindDir("tests")
+	fileReader, err := os.Open(filepath.Join(path, "testplugin.tar.gz"))
+	require.NoError(t, err)
+	defer fileReader.Close()
+
+	_, appErr := th.App.WriteFile(fileReader, th.App.getBundleStorePath("testplugin"))
+	checkNoError(t, appErr)
+
+	appErr = th.App.SyncPlugins()
+	checkNoError(t, appErr)
+
+	// Check if installed
+	pluginStatus, err := env.Statuses()
+	require.Nil(t, err)
+	require.Len(t, pluginStatus, 1)
+	require.Equal(t, "testplugin", pluginStatus[0].PluginId)
+
+	// No canary tag
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.PluginSettings.CanaryTag = ""
+	})
+	fileReader, err = os.Open(filepath.Join(path, "testplugin_v2.tar.gz"))
+	require.NoError(t, err)
+	defer fileReader.Close()
+
+	_, appErr = th.App.WriteFile(fileReader, th.App.getBundleStorePath("testplugin_v2"))
+	checkNoError(t, appErr)
+
+	appErr = th.App.SyncPlugins()
+	checkNoError(t, appErr)
+	// should be installed plugin with no canary flag
+	pluginStatus, err = env.Statuses()
+	require.Nil(t, err)
+	require.Len(t, pluginStatus, 1)
+	require.Equal(t, "testplugin_v2", pluginStatus[0].PluginId)
+
+	// with canary tag set
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.PluginSettings.CanaryTag = "daily"
+	})
+
+	appErr = th.App.SyncPlugins()
+	checkNoError(t, appErr)
+	// should be installed both plugins
+	pluginStatus, err = env.Statuses()
+	require.Nil(t, err)
+	require.Len(t, pluginStatus, 2)
```

Comment:
Given my comment above, this result surprises me -- after adding a canary tag, I'd have expected the same number of plugins installed -- just that the server would have installed something named `testplugin_v2.tar.gz+daily` -- but I don't see a test that sets up such a file?

Reply:
Here we are checking what happens if server has `daily` tag. In this case there are 2 plugins one with `+daily` ending and another one without ending. They both should show.